File: flutter/packages/hatake_core/lib/src/definition/sub_table_source.dart
import 'package:equatable/equatable.dart';

/// Where a `subTable`'s child rows come from when they are **not** embedded in
/// the parent record: their own repository, paged and linked by a foreign key.
///
/// Presence of a source changes the field's semantics (see `dsl-spec`): rows are
/// fetched with `search({parentKey: <parent key>}, page, pageSize)`, saved one at
/// a time, and the parent's `FormValidator` skips the field entirely because its
/// value does not live in the record.
class SubTableSource extends Equatable {
  /// Repository key for the child rows.
  final String repository;

  /// Child field holding the parent key. Passed as the search filter
  /// `{parentKey: <parent key value>}`.
  final String parentKey;

  /// **子の行を1件に指す項目**（定義に書いた順）。2つ以上なら複合キー。
  /// DSL キー: `key`。
  final List<String> keyFields;

  /// Rows per page.
  final int pageSize;

  const SubTableSource({
    required this.repository,
    required this.parentKey,
    this.keyFields = const ['id'],
    this.pageSize = 20,
  });

  @override
  List<Object?> get props => [repository, parentKey, keyFields, pageSize];
}

File: java/src/main/java/io/hatake/core/SubTableSource.java
package io.hatake.core;

import java.util.List;

/**
 * 明細（{@code type: subTable}）の子行を親レコードに埋め込まず、
 * <b>別 Repository から外部キーで引く</b>ときの取得元。DSL キーは {@code source}。
 *
 * <p>サーバ側での意味はひとつ: 子行は親レコードに入っていないので、
 * {@link FormValidator} はその項目を検証対象から外す。
 *
 * @param repository 子行の Repository キー
 * @param parentKey 子行が持つ親キーの項目名
 * @param keyFields 子行を1件に指す項目（定義に書いた順。DSL キー {@code key}、既定 {@code id}）。
 *     2つ以上なら複合キーで、並びは REST の道の順になる
 * @param pageSize 1 ページの行数（既定 20）
 */
public record SubTableSource(
        String repository,
        String parentKey,
        List<String> keyFields,
        int pageSize) {
}
